C-Plus-Plus/cp/Generate_all_valid_parentheses_combinations.cpp

```diff
@@ -0,0 +1,154 @@
+                                         /          \
+                                        /            \
+       output_string "((" {open = 0,close = 2}      output_string "()" {open = 1 , close = 1}
+                                \                                     /
+                                 \ choose )              choose (    /
+                                  \                                 /
+                                   \                               /
+          output_string "(()" {open = 0, close = 1}         output_string "()(" {open = 0 , close = 1}
+                                \                                      \
+                                 \ choose )                             \
+                                  \                                      \  choose )
+                                   \                                      \
+          output_string "(())" {open = 0, close = 0}           output_string "()()" {open = 0, close = 0}
+
+                             *STOP AND RETURN*                                       *STOP AND RETURN*
+
+
+
+*/
+/* Input  : an interger n
+   Output : vector of strings which will store our valid parentheses combination strings.
+*/
+
+// Code part
+
+#include<bits/stdc++.h>
+using namespace std;
+
+
+// This function is used to generate all well-forrmed combination of parentheses
+
+void generateParenthesis(int open, int close, string op, vector<string>&ans)
+{
+	// Base case : if count of open and close parentheses is both 0.
+
+	if (open == 0 && close == 0)
+	{
+		ans.push_back(op);  // push the output valid string combination to the answer vector and return
+
+		return;
+	}
+
+	if (open != 0)     // open bracket will always be one of the choice , given only the condition that open != 0
+	{
+		string op1 = op;
+
+		op1.push_back('(');  // output1 string (like we can see in recursive tree , it is the output string of when  we select '(')
+
+		// now , count of open brackets will decrease by 1 as we have used it now , and output string will be given as the op1 string now.
+		generateParenthesis(open - 1 , close, op1 , ans);
+	}
+
+  //close bracket will only be chooses if its count is not 0 , i.e close != 0 and count of close parenteses > count of open parenteses, i.e close > open
+
+  // As the closing parenthesis cannot occur before the open parenthesis for generating valid combination.
+
+	if (close != 0 && close > open)
+	{
+		string op2 = op;
+
+		op2.push_back(')');  // output2 string (like we can see in recursive tree , it is the output string of when  we select ')')
+
+        // now , count of close brackets will decrease by 1 as we have used it now , and output string will be given as the op2 string now.
+		generateParenthesis(open, close - 1, op2, ans);
+	}
+}
+
+int main()
+{
+	int n;
+
+	cout << "Enter the number:";
+
+	cin >> n;
+
+	vector<string>ans;  // this is our output vector
```

Comment:
Comments should be above the line of code. and not alongside. Please work on it. Rest everything is fine. Great work!